tsdb/engine/tsm1/compact_property_test.go

```diff
@@ -0,0 +1,173 @@
+	Name:        "No Gaps Rule",
+	Description: "Files should not have gaps within the same compaction level - if files A and C are in the same level group, any file B between them should also be in a group at the same level or higher",
+	Validator:   validateNoGaps,
+}
+
+// parseFileName uses the existing TSM file parsing functionality
+func parseFileName(filename string) (generation int, sequence int, err error) {
+	return tsm1.DefaultParseFileName(filename)
+}
+
+// fileInfo holds parsed file information for sorting
+type fileInfo struct {
+	filename   string
+	generation int
+	sequence   int
+}
+
+// validateNoGaps checks that there are no gaps within the same compaction level
+// A gap occurs when files A and C are in the same level group, but file B (between A and C)
+// is in a different level group, creating a "hole" in the sequence
+func validateNoGaps(allFiles []string, groups []tsm1.CompactionGroup) error {
+	// Parse all files and sort them by generation, then by sequence
+	var fileInfos []fileInfo
+	for _, file := range allFiles {
+		gen, seq, err := parseFileName(file)
+		if err != nil {
+			return fmt.Errorf("failed to parse file %s: %v", file, err)
+		}
+		fileInfos = append(fileInfos, fileInfo{
+			filename:   file,
+			generation: gen,
+			sequence:   seq,
+		})
+	}
+
+	// Sort by generation first, then by sequence
+	sort.Slice(fileInfos, func(i, j int) bool {
+		if fileInfos[i].generation != fileInfos[j].generation {
+			return fileInfos[i].generation < fileInfos[j].generation
+		}
+		return fileInfos[i].sequence < fileInfos[j].sequence
+	})
+
+	// Create a map of filename to group index
+	fileToGroup := make(map[string]int)
+	for groupIdx, group := range groups {
+		for _, file := range group {
+			fileToGroup[file] = groupIdx
+		}
+	}
+
+	// Group files by their compaction group
+	groupToFiles := make(map[int][]fileInfo)
+	for _, info := range fileInfos {
+		if groupIdx, inGroup := fileToGroup[info.filename]; inGroup {
+			groupToFiles[groupIdx] = append(groupToFiles[groupIdx], info)
+		}
+	}
+
+	// For each group, check if there are gaps in the file sequence
+	for groupIdx, filesInGroup := range groupToFiles {
+		if len(filesInGroup) < 2 {
+			continue // No gaps possible with less than 2 files
+		}
+
+		// Check for gaps between files in this group
+		for i := 0; i < len(filesInGroup)-1; i++ {
+			current := filesInGroup[i]
+			next := filesInGroup[i+1]
+
+			// Find all files between current and next in the sorted sequence
+			currentPos := findFilePosition(current, fileInfos)
+			nextPos := findFilePosition(next, fileInfos)
+
+			// Check if there are files between current and next that are in different groups
+			for pos := currentPos + 1; pos < nextPos; pos++ {
+				betweenFile := fileInfos[pos]
+				betweenGroup, inGroup := fileToGroup[betweenFile.filename]
+
+				if inGroup && betweenGroup != groupIdx {
+					return fmt.Errorf("gap detected: files %s and %s are in group %d, but file %s (between them) is in group %d",
```

Comment:
`expected results: No Gaps Rule violation: gap detected: files 04-04.tsm and 09-04.tsm are in group 1, but file 05-03.tsm (between them) is in group 0`

This error message is tough as the group indices don't tell me anything about which level group it came from. I think it should know which level group its talking about. 

also: "no gaps rule" is odd description of the property. Maybe its me: the rule is that tsm files within a group must be adjacent in the master list of files. "Adjacency rule" seems more accurate. a "gap" sounds like a _missing_ tsm file instead of one in the wrong place.

Reply:
also, it'd be nice if all instances of the issues were reported instead of stopping at the first adjacency rule violation.